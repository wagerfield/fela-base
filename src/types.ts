import { BorderRuleProps } from "./rules/border/border"
import { ButtonRuleProps } from "./rules/button/button"
import { DiscRuleProps } from "./rules/disc/disc"
import { DisplayRuleProps } from "./rules/display/display"
import { EdgeRuleProps } from "./rules/edge/edge"
import { EllipsisRuleProps } from "./rules/ellipsis/ellipsis"
import { FillRuleProps } from "./rules/fill/fill"
import { FlexRuleProps } from "./rules/flex/flex"
import { LayoutRuleProps } from "./rules/layout/layout"
import { ListRuleProps } from "./rules/list/list"
import { OutlineRuleProps } from "./rules/outline/outline"
import { ScrollRuleProps } from "./rules/scroll/scroll"
import { ShadowRuleProps } from "./rules/shadow/shadow"
import { TextRuleProps } from "./rules/text/text"
import { TransitionRuleProps } from "./rules/transition/transition"

// Edge

export type EdgeKey = "top" | "right" | "bottom" | "left"

export type EdgeMap = Partial<Record<EdgeKey, any>>

export type EdgeSet = EdgeKey[]

export type Edge = EdgeKey | EdgeMap | EdgeSet

// Ease

export type EaseKey = "IN" | "OUT" | "IN_OUT"

export type EaseMap = Record<EaseKey, string>

// Fela

export interface Props {
  [key: string]: any
}

export interface Style {
  [key: string]: any
}

export interface RuleDefaults {
  border?: BorderRuleProps
  button?: ButtonRuleProps
  disc?: DiscRuleProps
  display?: DisplayRuleProps
  edge?: EdgeRuleProps
  ellipsis?: EllipsisRuleProps
  fill?: FillRuleProps
  flex?: FlexRuleProps
  layout?: LayoutRuleProps
  list?: ListRuleProps
  outline?: OutlineRuleProps
  scroll?: ScrollRuleProps
  shadow?: ShadowRuleProps
  text?: TextRuleProps
  transition?: TransitionRuleProps
}

export type RuleName = keyof RuleDefaults

export type Rule = (props: Props) => Style

export type Unit = number | string

// Config

export interface Aliases {
  [name: string]: string
}

export interface Breakpoints {
  [name: string]: number
}

export interface CustomProps {
  [key: string]: (value: any) => Style
}
