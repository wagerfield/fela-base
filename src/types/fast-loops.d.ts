declare module "fast-loops" {
  export function arrayReduce<T, U extends any[]>(
    arr: U,
    reducer: (
      accumulator: T,
      value: any,
      index: number,
      length: number,
      arr: U
    ) => T,
    accumulator: T
  ): T

  export function objectReduce<T, U extends object>(
    obj: U,
    reducer: (
      accumulator: T,
      value: any,
      key: string,
      obj: U
    ) => T,
    accumulator: T
  ): T
}
