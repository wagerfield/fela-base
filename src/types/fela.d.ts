declare module "fela-plugin-embedded" {
  import { TPlugin } from "fela"
  export default function(): TPlugin
}

declare module "fela-plugin-named-keys" {
  import { TPlugin } from "fela"
  export interface Config {
    [alias: string]: string
  }
  export default function(config: Config): TPlugin
}

declare module "fela-plugin-custom-property" {
  import { TPlugin } from "fela"
  export interface Config {
    [key: string]: (value: any) => object
  }
  export default function(config: Config): TPlugin
}
