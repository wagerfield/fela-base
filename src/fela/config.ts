import { IConfig as FelaConfig } from "fela"
import monolithic from "fela-monolithic"
import extendPlugin from "fela-plugin-extend"
import embeddedPlugin from "fela-plugin-embedded"
import fallbackPlugin from "fela-plugin-fallback-value"
import pseudoPlugin from "fela-plugin-friendly-pseudo-class"
import placeholderPlugin from "fela-plugin-placeholder-prefixer"
import prefixerPlugin from "fela-plugin-prefixer"
import lvhaPlugin from "fela-plugin-lvha"
import unitPlugin, { Unit, UnitPerProperty } from "fela-plugin-unit"
import { Aliases, Breakpoints, CustomProps, RuleDefaults } from "../types"
import { setRuleDefaults } from "../rules/defaults"
import customPropsPlugin from "./custom-props"
import namedKeysPlugin from "./named-keys"

export interface UnitOptions {
  unitMap: UnitPerProperty
  defaultUnit: Unit
}

export interface ConfigOptions {
  useMonolithicClasses?: boolean
  usePrettySelectors?: boolean
  ruleDefaults?: RuleDefaults
  customProps?: CustomProps
  breakpoints?: Breakpoints
  aliases?: Aliases
  units?: UnitOptions
}

export default ({
  useMonolithicClasses = false,
  usePrettySelectors = false,
  ruleDefaults,
  customProps,
  breakpoints,
  aliases,
  units
}: ConfigOptions = {}): FelaConfig => {
  // Set rule defaults
  if (ruleDefaults) setRuleDefaults(ruleDefaults)

  // Add monolithic enhancer
  const enhancers = useMonolithicClasses
    ? [monolithic({ prettySelectors: usePrettySelectors })]
    : []

  // Unit plugin values
  const defaultUnit = (units && units.defaultUnit) || "px"
  const unitMap = (units && units.unitMap) || {}

  // Add plugins: order matters
  const plugins = [
    extendPlugin(),
    pseudoPlugin(),
    namedKeysPlugin({ aliases, breakpoints }),
    customPropsPlugin(customProps),
    placeholderPlugin(),
    embeddedPlugin(),
    prefixerPlugin(),
    fallbackPlugin(),
    lvhaPlugin(),
    unitPlugin(defaultUnit, unitMap)
  ]

  // Return config
  return { enhancers, plugins }
}
