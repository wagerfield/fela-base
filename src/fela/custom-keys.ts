import { TPlugin as FelaPlugin } from "fela"
import customKeys, {
  Config as CustomKeys
} from "fela-plugin-custom-property"
import { ALL } from "../core/edges"
import { isObject } from "../core/utils"
import ellipsis from "../rules/ellipsis/ellipsis"
import border from "../rules/border/border"
import edges from "../rules/edges/edges"

export {
  Config as CustomKeys
} from "fela-plugin-custom-property"

export default (options: CustomKeys = {}): FelaPlugin =>
  customKeys({
    border: (value) =>
      isObject(value) ? border(value) : { border: value },

    ellipsis: (value) => (value ? ellipsis() : {}),

    fontFeatures: (value) => ({
      "-webkit-font-feature-settings": value,
      "font-feature-settings": value
    }),

    fontSmoothing: (value) => ({
      "-webkit-font-smoothing": value,
      "-moz-osx-font-smoothing": "grayscale"
    }),

    margin: (value) =>
      edges({
        prefix: "margin",
        edges: ALL,
        value
      }),

    padding: (value) =>
      edges({
        prefix: "padding",
        edges: ALL,
        value
      }),

    scrolling: (value) => ({
      "-webkit-overflow-scrolling": value
    }),

    size: (value) => ({
      width: value,
      height: value
    }),

    tapColor: (value) => ({
      "-webkit-tap-highlight-color": value
    }),

    ...options
  })
