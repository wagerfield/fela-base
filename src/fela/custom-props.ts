import { TPlugin as FelaPlugin } from "fela"
import customKeys from "fela-plugin-custom-property"
import { background, border, edge, ellipsis, outline } from "../rules"
import { CustomProps, Rule } from "../types"
import { isObject } from "../core/utils"
import { ALL } from "../core/edges"

const customProp = (rule: Rule, key: string) => (value: any) =>
  isObject(value) ? rule(value) : { [key]: value }

const customEdge = (key: string) => (value: any) =>
  edge({
    edge: isObject(value) ? value : ALL,
    prefix: key,
    value
  })

export default (props?: CustomProps): FelaPlugin =>
  customKeys({
    background: customProp(background, "background"),

    border: customProp(border, "border"),

    outline: customProp(outline, "outline"),

    margin: customEdge("margin"),

    padding: customEdge("padding"),

    ellipsis: (value) =>
      isObject(value) ? ellipsis(value) : value === true ? ellipsis() : {},

    fontFeatures: (value) => ({
      "-webkit-font-feature-settings": value,
      "font-feature-settings": value
    }),

    fontSmoothing: (value) => ({
      "-webkit-font-smoothing": value,
      "-moz-osx-font-smoothing": "grayscale"
    }),

    scrolling: (value) => ({
      "-webkit-overflow-scrolling": value
    }),

    size: (value) => ({
      width: value,
      height: value
    }),

    tapColor: (value) => ({
      "-webkit-tap-highlight-color": value
    }),

    ...props
  })
