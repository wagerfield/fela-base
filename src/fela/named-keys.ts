import { objectReduce } from "fast-loops"
import { TPlugin as FelaPlugin } from "fela"
import namedKeys from "fela-plugin-named-keys"

import { kebabToCamel } from "../core/utils"

export interface Breakpoints {
  [name: string]: number
}

export interface Aliases {
  [name: string]: string
}

export interface NamedKeysOptions {
  aliases?: Aliases
  breakpoints?: Breakpoints
}

export const buildMediaQuery = (
  value: number,
  below: boolean
) => {
  const prefix = below ? "max" : "min"
  const offset = below ? value : value + 1
  return `@media (${prefix}-width: ${offset}px)`
}

export const buildMediaQueries = (
  breakpoints: Breakpoints
) => {
  const acc: Aliases = {}
  return objectReduce(
    breakpoints,
    (result, value, key) => {
      const belowKey = kebabToCamel(`below-${key}`)
      const aboveKey = kebabToCamel(`above-${key}`)
      result[belowKey] = buildMediaQuery(value, true)
      result[aboveKey] = buildMediaQuery(value, false)
      return result
    },
    acc
  )
}

export default ({
  aliases = {},
  breakpoints = {}
}: NamedKeysOptions = {}): FelaPlugin =>
  namedKeys({
    placeholder: "::placeholder",
    ...buildMediaQueries(breakpoints),
    ...aliases
  })
