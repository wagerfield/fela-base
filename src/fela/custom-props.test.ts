import { createRenderer, IRenderer as FelaRenderer } from "fela"
import { CustomProps, Style } from "../types"
import customProps from "./custom-props"

const snapshotStyle = (name?: string) => (style) => {
  expect(style).toMatchSnapshot(name)
  return style
}

const createTestRenderer = (props?: CustomProps) => {
  return createRenderer({
    plugins: [customProps(props), snapshotStyle()]
  })
}

const renderStyle = (renderer: FelaRenderer, style: Style) => {
  renderer.renderRule(() => style, {})
}

describe("customProps(props)", () => {
  const renderer = createTestRenderer()

  it("returns a function", () => {
    expect(customProps()).toEqual(expect.any(Function))
  })

  it("adds custom props", () => {
    const customRenderer = createTestRenderer({
      customProp: (value) => ({
        customPropA: value,
        customPropB: value
      })
    })
    renderStyle(customRenderer, {
      customProp: "duplicate"
    })
  })

  it("renders background value", () => {
    renderStyle(renderer, {
      background: "red"
    })
  })

  it("renders background object", () => {
    renderStyle(renderer, {
      background: {
        color: "blue",
        repeat: "no-repeat"
      }
    })
  })

  it("renders border value", () => {
    renderStyle(renderer, {
      border: "none"
    })
  })

  it("renders border object", () => {
    renderStyle(renderer, {
      border: {
        edge: ["left", "right"],
        width: "10px",
        color: "blue"
      }
    })
  })

  it("renders outline value", () => {
    renderStyle(renderer, {
      outline: "none"
    })
  })

  it("renders outline object", () => {
    renderStyle(renderer, {
      outline: {
        width: "10px",
        color: "blue"
      }
    })
  })

  it("renders default ellipsis object", () => {
    renderStyle(renderer, {
      ellipsis: true
    })
  })

  it("renders custom ellipsis object", () => {
    renderStyle(renderer, {
      ellipsis: {
        overflow: "visible"
      }
    })
  })

  it("does not render ellipsis object", () => {
    renderStyle(renderer, {
      ellipsis: false
    })
  })

  it("renders font feature object", () => {
    renderStyle(renderer, {
      fontFeatures: "liga"
    })
  })

  it("renders font smoothing object", () => {
    renderStyle(renderer, {
      fontSmoothing: "antialiased"
    })
  })

  it("renders margin value", () => {
    renderStyle(renderer, {
      margin: 10
    })
  })

  it("renders margin object", () => {
    renderStyle(renderer, {
      margin: {
        right: 20,
        top: 10
      }
    })
  })

  it("renders padding value", () => {
    renderStyle(renderer, {
      padding: 10
    })
  })

  it("renders padding object", () => {
    renderStyle(renderer, {
      padding: {
        right: 20,
        top: 10
      }
    })
  })

  it("renders scrolling object", () => {
    renderStyle(renderer, {
      scrolling: "touch"
    })
  })

  it("renders size object", () => {
    renderStyle(renderer, {
      size: 10
    })
  })

  it("renders tap color object", () => {
    renderStyle(renderer, {
      tapColor: "red"
    })
  })
})
