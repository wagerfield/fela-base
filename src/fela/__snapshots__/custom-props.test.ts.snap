// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customProps(props) adds custom props 1`] = `
Object {
  "customPropA": "duplicate",
  "customPropB": "duplicate",
}
`;

exports[`customProps(props) does not render ellipsis object 1`] = `Object {}`;

exports[`customProps(props) renders background object 1`] = `
Object {
  "backgroundAttachment": undefined,
  "backgroundClip": undefined,
  "backgroundColor": "blue",
  "backgroundImage": undefined,
  "backgroundOrigin": undefined,
  "backgroundPosition": undefined,
  "backgroundPositionX": undefined,
  "backgroundPositionY": undefined,
  "backgroundRepeat": "no-repeat",
  "backgroundSize": undefined,
}
`;

exports[`customProps(props) renders background value 1`] = `
Object {
  "background": "red",
}
`;

exports[`customProps(props) renders border object 1`] = `
Object {
  "borderLeftColor": "blue",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "10px",
  "borderRightColor": "blue",
  "borderRightStyle": "solid",
  "borderRightWidth": "10px",
}
`;

exports[`customProps(props) renders border value 1`] = `
Object {
  "border": "none",
}
`;

exports[`customProps(props) renders custom ellipsis object 1`] = `
Object {
  "overflow": "visible",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
}
`;

exports[`customProps(props) renders default ellipsis object 1`] = `
Object {
  "overflow": "hidden",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
}
`;

exports[`customProps(props) renders font feature object 1`] = `
Object {
  "-webkit-font-feature-settings": "liga",
  "font-feature-settings": "liga",
}
`;

exports[`customProps(props) renders font smoothing object 1`] = `
Object {
  "-moz-osx-font-smoothing": "grayscale",
  "-webkit-font-smoothing": "antialiased",
}
`;

exports[`customProps(props) renders margin object 1`] = `
Object {
  "marginRight": 20,
  "marginTop": 10,
}
`;

exports[`customProps(props) renders margin value 1`] = `
Object {
  "marginBottom": 10,
  "marginLeft": 10,
  "marginRight": 10,
  "marginTop": 10,
}
`;

exports[`customProps(props) renders outline object 1`] = `
Object {
  "outlineColor": "blue",
  "outlineOffset": undefined,
  "outlineStyle": undefined,
  "outlineWidth": "10px",
}
`;

exports[`customProps(props) renders outline value 1`] = `
Object {
  "outline": "none",
}
`;

exports[`customProps(props) renders padding object 1`] = `
Object {
  "paddingRight": 20,
  "paddingTop": 10,
}
`;

exports[`customProps(props) renders padding value 1`] = `
Object {
  "paddingBottom": 10,
  "paddingLeft": 10,
  "paddingRight": 10,
  "paddingTop": 10,
}
`;

exports[`customProps(props) renders scrolling object 1`] = `
Object {
  "-webkit-overflow-scrolling": "touch",
}
`;

exports[`customProps(props) renders size object 1`] = `
Object {
  "height": 10,
  "width": 10,
}
`;

exports[`customProps(props) renders tap color object 1`] = `
Object {
  "-webkit-tap-highlight-color": "red",
}
`;
